gobject-introspection: support cross-compilation

Artturin opened the pull request:
with this json-glib cross-compiles and the gir files exist, have not tested running any binaries on aarch64
```
nix build ".#pkgsCross.aarch64-multiplatform.json-glib"
```


https://github.com/NixOS/nixpkgs/pull/88222
https://github.com/NixOS/nixpkgs/issues/72868

https://github.com/void-linux/void-packages/blob/master/srcpkgs/gobject-introspection/template and the buildroot mk file were inspirations

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Mindavi commented:
Hmm, will we need to build such wrapper for each package using gobject-introspection?

SuperSandro2000 commented on pkgs/development/libraries/json-glib/default.nix:
```diff
@@ -6,19 +6,46 @@
+  gobject-introspection-wrapper = runCommand "make-wrappers" {
+    depsBuildBuild = [ pkgsBuildBuild.makeBinaryWrapper ];
+  } ''
+    makeWrapper ${buildPackages.gobject-introspection.dev}/bin/g-ir-scanner \
```
```suggestion
    makeWrapper ${lib.getDev buildPackages.gobject-introspection}/bin/g-ir-scanner \
```
Always nice to use the helper function in case dev output gets removed.

SuperSandro2000 commented on pkgs/development/tools/misc/prelink/default.nix:
```diff
@@ -1,20 +1,37 @@
-    url = "https://people.redhat.com/jakub/prelink/prelink-${version}.tar.bz2";
-    sha256 = "1w20f6ilqrz8ca51qhrn1n13h7q1r34k09g33d6l2vwvbrhcffb3";
-  };
+  # There are some failures to investigate
```
```suggestion
  # TODO: There are some failures to investigate
```

SuperSandro2000 commented on pkgs/development/tools/misc/prelink/default.nix:
```diff
@@ -1,20 +1,37 @@
 stdenv.mkDerivation rec {
   pname = "prelink";
-  version = "20130503";
+  version = "2019-6-24-" + lib.substring 0 7 src.rev;
```
```suggestion
  version = "unstable-2019-6-24";
```

SuperSandro2000 commented on pkgs/development/tools/misc/prelink/default.nix:
```diff
@@ -1,20 +1,37 @@
   meta = {
-    homepage = "https://people.redhat.com/jakub/prelink/";
+    homepage = "https://wiki.yoctoproject.org/wiki/Cross-Prelink";
+    #homepage = "https://people.redhat.com/jakub/prelink/";
```
Thats just the old download page

```suggestion
```

Artturin replied:
for now i just copied the prelink from ericson's pr. i'll clean it up later

SuperSandro2000 commented on pkgs/development/libraries/json-glib/default.nix:
```diff
@@ -37,23 +64,29 @@ stdenv.mkDerivation rec {
     glib
+    gi-docgen
+    libxslt
+    (if (stdenv.buildPlatform == stdenv.hostPlatform) then gobject-introspection else gobject-introspection-wrapper)
```
If that is something universal we need everywhere we should move it into top-level or to the cross compile specific overlays.

SuperSandro2000 commented on pkgs/development/libraries/json-glib/default.nix:
```diff
@@ -6,19 +6,46 @@
+      $out/bin/g-ir-compiler \
+      --inherit-argv0 \
+      --add-flags '${gobject-introspection.dev}/bin/g-ir-compiler'
+
+    #shopt -s extglob dotglob
+    #cp -r ${gobject-introspection}/!(bin) -t $out
+    #cp -r ${gobject-introspection.dev}/!(bin) -t $out
+    #shopt -u extglob dotglob
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/json-glib/default.nix:
```diff
@@ -6,19 +6,46 @@
+  } ''
+    makeWrapper ${buildPackages.gobject-introspection.dev}/bin/g-ir-scanner \
+      $out/bin/g-ir-scanner \
+      --inherit-argv0 \
+      --add-flags '--use-binary-wrapper=${stdenv.hostPlatform.emulator buildPackages}' \
+      --add-flags '--use-ldd-wrapper=${buildPackages.prelink}/bin/prelink-rtld' \
+      --add-flags '--add-include-path=${buildPackages.gobject-introspection.dev}/share/gir-1.0'
```
```suggestion
        --inherit-argv0 \
        --add-flags '--use-binary-wrapper=${stdenv.hostPlatform.emulator buildPackages}' \
        --add-flags '--use-ldd-wrapper=${buildPackages.prelink}/bin/prelink-rtld' \
        --add-flags '--add-include-path=${buildPackages.gobject-introspection.dev}/share/gir-1.0'
```

SuperSandro2000 commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -56,6 +58,8 @@ stdenv.mkDerivation rec {
     })
   ];
 
+  #depsBuildBuild = [ buildPackages.stdenv.cc ];
+
```
```suggestion
```

SuperSandro2000 reviewed (commented):
Only some formatting improvements. Thanks for trying to bring this forward. I really really appreciate it!

Artturin commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -119,6 +134,32 @@ stdenv.mkDerivation rec {
+        --inherit-argv0 \
+        --add-flags '--use-binary-wrapper=${stdenv.hostPlatform.emulator buildPackages}' \
+        --add-flags '--use-ldd-wrapper=${buildPackages.prelink}/bin/prelink-rtld' \
+        --add-flags '--add-include-path=${buildPackages.gobject-introspection.dev}/share/gir-1.0'
```
this may not be necessary for us

Mindavi commented:
I do like this better than having it in json-glib at least :). Still need to try it out

Artturin committed:
meson: add mesonEmulatorHook

fixes building documentation while cross-compiling and other issues
Exec format error: './gdk3-scan'

added some simple documentation

Artturin committed:
prelink: 20130503 -> unstable-2019-06-24 cross_prelink branch

has prelink-rltd and the normal prelink tools

Artturin committed:
gobject-introspection: support cross-compilation

used the following as references

https://github.com/void-linux/void-packages/blob/master/srcpkgs/gobject-introspection

and

https://git.busybox.net/buildroot/tree/package/gobject-introspection

thanks void and buildroot

Mindavi commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -92,7 +95,24 @@ python3.pkgs.buildPythonApplication rec {
+        [binaries]
+        # fixes errors such as
+        # [Errno 8] Exec format error: './gdk3-scan'
+        exe_wrapper = ${lib.escapeShellArg (stdenv.targetPlatform.emulator buildPackages)}
```
This maybe creates a big loop when you want to cross-compile a meson package and it's somehow a dependency of qemu. Then you need to rebuild hundreds of packages when you're just changing 1 file in a bad case :(. Not sure if we can easily prevent that, but that's really annoying if you're trying to fix something simple.

Artturin replied:
It could be a separate hook that's added when needed

SuperSandro2000 replied:
I agree with Mindavi that if it is easily possible to move this to its own hook we should do that. I would imagine doing something similar like the CPython package does https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/cpython/default.nix#L367= 

I think we should also add it to passthru in case we might want to use the hook somewhere else.

Artturin replied:
done

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17565,7 +17565,9 @@ with pkgs;
   gns3-server = gns3Packages.serverStable;
 
-  gobject-introspection = callPackage ../development/libraries/gobject-introspection {
+  gobject-introspection = if (stdenv.hostPlatform != stdenv.targetPlatform) then callPackage ../development/libraries/gobject-introspection/wrapper.nix { } else gobject-introspection-unwrapped;
+
+  gobject-introspection-unwrapped = callPackage ../development/libraries/gobject-introspection {
```
```suggestion
  gobject-introspection = if (stdenv.hostPlatform != stdenv.targetPlatform) then gobject-introspection-wrapped else gobject-introspection-unwrapped;
  
  gobject-introspection-wrapped = callPackage ../development/libraries/gobject-introspection/wrapper.nix { }

  gobject-introspection-unwrapped = callPackage ../development/libraries/gobject-introspection {
```
If we ever explicitly need the wrapped or unwrapped version we could do this to make issues easier to debug.

Artturin replied:
i don't think it it should be exposed like that, it causes a eval error and its not usable

```
$ nix build ".#pkgsCross.aarch64-multiplatform.gobject-introspection-wrapped"
error: attribute 'gobject-introspection-unwrapped' missing

       at /nix/store/cza07hhvf1n83y9c4ar2dpvx10rbziah-source/pkgs/development/libraries/gobject-introspection/wrapper.nix:25:38:

           24|
           25|       export targetgir="${lib.getDev targetPackages.gobject-introspection-unwrapped}"
             |                                      ^
           26|
(use '--show-trace' to show detailed location information)
```

i've added a comment for how to build it to wrapper.nix

SuperSandro2000 commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -92,7 +95,24 @@ python3.pkgs.buildPythonApplication rec {
+    let
+      crossFile = writeText "cross-file.conf" ''
+        [binaries]
+        # fixes errors such as
+        # [Errno 8] Exec format error: './gdk3-scan'
```
I think we should change the comment to mention that we want to choose the binary for the right platform when cross compiling.

SuperSandro2000 reviewed (commented):
I think we are in a good state here now!

@Mindavi what do you think? Can we merge this soon?

Artturin committed:
gobject-introspection: add artturin as maintainer

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17565,7 +17565,10 @@ with pkgs;
   gns3-server = gns3Packages.serverStable;
 
-  gobject-introspection = callPackage ../development/libraries/gobject-introspection {
+  gobject-introspection = if (stdenv.hostPlatform != stdenv.targetPlatform)
```
Is this a good place for something like this, as opposed to e.g. https://github.com/NixOS/nixpkgs/blob/a785565fb9f75c7361232b3da1c4384e8f0da8a2/pkgs/stdenv/cross/default.nix?

jtojnar commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -101,6 +105,28 @@ python3.pkgs.buildPythonApplication rec {
+
+    if (stdenv.buildPlatform != stdenv.hostPlatform) then
+      makeSetupHook { name = "mesonCrossEmulatorWrapper"; } crossHook
+    else "";
```
Maybe make this `throw` so that people do not accidentally introduce empty string to inputs, which could be an issue if we ever decide to be more strict about inputs (cannot find the relevant PR now).

Artturin replied:
done and added a example to the docs

Artturin commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -101,6 +105,28 @@ python3.pkgs.buildPythonApplication rec {
+      '';
+      crossHook = writeText "crossHook" ''
+        add_meson_exe_wrapper_cross_flag() {
+          mesonFlagsArray+=(--cross-file=${crossFile})
```
`--cross-file` can be specified multiple times, newest invocations will override the older ones 
(we actually specify a cross-file in make-derivation.nix already)

Mindavi replied:
The other cross file is for packages that want to link to llvm IIRC, in theory a package could want both gir and llvm

Artturin commented:
Remember that gobject-introspection has to be in both buildInputs and nativeBuildInputs

Artturin commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -40,6 +42,9 @@ stdenv.mkDerivation rec {
   patches = [
+    # prelink-rtld, which we use for cross returns 127 when it can't find a library.
+    # https://git.busybox.net/buildroot/tree/package/gobject-introspection/0003-giscanner-ignore-error-return-codes-from-ldd-wrapper.patch
+    ./giscanner-ignore-error-return-codes-from-ldd-wrapper.patch
```
i decided to inline the patch so i can easily update it if needed

Artturin commented:
lets do the removing of `enableIntrospection`'s in a separate pr

Ericson2314 commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -101,6 +105,28 @@ python3.pkgs.buildPythonApplication rec {
+  # which needs an emulator
+  # example of an error which this fixes
+  # [Errno 8] Exec format error: './gdk3-scan'
+  passthru.crossEmulatorWrapper =
```
I don't really like mashing things in `passthru`. How do you all feel about putting this in `pkgs/build-support/setup-hooks`?

Artturin replied:
i moved the hook to all-packages.nix but i put the file in the meson dir instead of setup-hooks

Ericson2314 commented on pkgs/development/libraries/gobject-introspection/wrapper.nix:
```diff
@@ -0,0 +1,35 @@
+
+# to build, run
+# `nix build ".#pkgsCross.aarch64-multiplatform.buildPackages.gobject-introspection"`
+gobject-introspection-unwrapped.overrideAttrs (_previousAttrs: {
```
We're overriding? Can we make a fresh derivation like the other wrappers?

Artturin replied:
I tried that but dirs would have to be copied to the wrapper and paths rewritten and then split again, so this is the simplest and most maintainable way.

Ericson2314 commented on pkgs/development/libraries/gobject-introspection/wrappers/g-ir-compiler.sh:
```diff
@@ -0,0 +1,9 @@
+#! @bash@
+# shellcheck shell=bash
+
+# try to check if we're in a nix build
+if [ -n "${name:-}" ]; then
```
What is this about? I don't feel good about this.

Ericson2314 replied:
Other wrappers also don't work very well outside a Nix build, granted, but such special casing I don't think is the right way to fix that.

Mindavi commented on pkgs/development/libraries/gobject-introspection/giscanner-ignore-error-return-codes-from-ldd-wrapper.patch:
```diff
@@ -0,0 +1,30 @@
+         else:
+             args.extend(['ldd', binary.args[0]])
+-        output = subprocess.check_output(args)
++        output = subprocess.run(args, check=False, stdout=subprocess.PIPE).stdout
```
Meh, this is not really nice since it may also ignore any _real_ errors too :/

Mindavi replied:
Could we maybe only ignore status 127?

Artturin replied:
idk if we even need this patch, lets try running without it for a while?

Mindavi commented on pkgs/development/tools/misc/prelink/default.nix:
```diff
@@ -1,22 +1,46 @@
+  # PASS 36 FAIL 10
+  # ld: cannot find reloc2lib2.so: No such file or directory
+  # cp: cannot create regular file 'deps1.tree/lib/libstdc++.so.6.0.29.orig': Permission denied
+  doCheck = false;
```
Remove some testcases so we at least have some coverage?

Artturin replied:
done

Mindavi commented:
> I think we are in a good state here now!
> 
> @Mindavi what do you think? Can we merge this soon?

I think we can merge this soon-ish. I want to do some testing, but I tried out an older version (from yesterday) and that was already looking promising. Let's also address the review comments :smile:.

Just to get some idea of the status:

1. For any package _not_ using meson, we don't need to add the hook, right? It should work out of the box?
2. For any package using meson we need the hook because meson does it's own handing of gobject-introspection, correct?

I'm not really sure if we really want to make a separate hook or if we always want to do this when meson and gobject-introspection are both used. I think it would be more convenient and cause less friction if we could automatically add the cross-file when needed instead of having to explicitly define it.

In general, most of this is looking good!

Artturin commented:
> > I think we are in a good state here now!
> > @Mindavi what do you think? Can we merge this soon?
> 
> I think we can merge this soon-ish. I want to do some testing, but I tried out an older version (from yesterday) and that was already looking promising. Let's also address the review comments 😄.
> 
> Just to get some idea of the status:
> 
>     1. For any package _not_ using meson, we don't need to add the hook, right? It should work out of the box?
> 
>     2. For any package using meson we need the hook because meson does it's own handing of gobject-introspection, correct?
> 
> 
> I'm not really sure if we really want to make a separate hook or if we always want to do this when meson and gobject-introspection are both used. I think it would be more convenient and cause less friction if we could automatically add the cross-file when needed instead of having to explicitly define it.
> 
> In general, most of this is looking good!

The hook is not always needed even with meson, small example
json-glib doesn't need it while gtk3 does

Artturin committed:
gobject-introspection: revert patch to ignore return codes from ldd

we may not need this so reverting for now

Mindavi commented on pkgs/development/tools/misc/prelink/default.nix:
```diff
@@ -1,22 +1,52 @@
-    sha256 = "1w20f6ilqrz8ca51qhrn1n13h7q1r34k09g33d6l2vwvbrhcffb3";
-  };
+  preCheck = ''
+    for f in reloc2 layout1 unprel1 tls3 cxx2 cxx3 quick1 quick2 deps1 deps2; do
```
# Disable some tests because they're failing

Mindavi replied:
https://github.com/Gottox/prelink-cross

I see the project itself already is aware of some tests failing 😜

SuperSandro2000 replied:
Thats only a mirror. Real upstream is https://git.yoctoproject.org/prelink-cross/

Artturin replied:
done

SuperSandro2000 commented on pkgs/development/tools/build-managers/meson/default.nix:
```diff
@@ -1,10 +1,14 @@
 { lib
+, stdenv
 , fetchpatch
 , installShellFiles
 , ninja
 , pkg-config
 , python3
 , substituteAll
+, writeText
+, buildPackages
+, makeSetupHook
```
I think those are now unused.

SuperSandro2000 commented on pkgs/development/libraries/gobject-introspection/wrapper.nix:
```diff
@@ -0,0 +1,35 @@
+{ lib
+, stdenv
+, makeWrapper
+, buildEnv
+, writeText
+, runCommand
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/gobject-introspection/wrapper.nix:
```diff
@@ -0,0 +1,35 @@
+, runCommand
+, buildPackages
+, gobject-introspection-unwrapped
+, pkgsHostHost
+, pkgsBuildBuild
```
```suggestion
```

Mindavi reviewed (approved):
Diff now LGTM

Artturin commented:
Thanks everyone!
